// snyk-request-manager/dist/customErrors/apiError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenericError = exports.NotFoundError = exports.ApiAuthenticationError = exports.ApiError = void 0;
class ApiError extends Error {
    constructor(message) {
        var _a;
        super(message);
        this.name = "ApiError";
        this.message = (message || "");
        this.data = (((_a = message.response) === null || _a === void 0 ? void 0 : _a.data) || "");
    }
}
exports.ApiError = ApiError;
class ApiAuthenticationError extends Error {
    constructor(message) {
        var _a;
        super(message);
        this.name = "ApiAuthenticationError";
        this.message = (message || "");
        this.data = (((_a = message.response) === null || _a === void 0 ? void 0 : _a.data) || "");
    }
}
exports.ApiAuthenticationError = ApiAuthenticationError;
class NotFoundError extends Error {
    constructor(message) {
        var _a;
        super(message);
        this.name = "NotFoundError";
        this.message = (message || "");
        this.data = (((_a = message.response) === null || _a === void 0 ? void 0 : _a.data) || "");
    }
}
exports.NotFoundError = NotFoundError;
class GenericError extends Error {
    constructor(message) {
        var _a;
        super(message);
        this.name = "Unknown";
        this.message = (message || "");
        this.data = (((_a = message.response) === null || _a === void 0 ? void 0 : _a.data) || "");
    }
}
exports.GenericError = GenericError;
