// snyk-api-ts-client/dist/client/generated/orgs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Orgs = void 0;
const clientError_1 = require("../errors/clientError");
const snyk_request_manager_1 = require("snyk-request-manager");
const requestManager = new snyk_request_manager_1.requestsManager({
    burstSize: 100,
    period: 100,
    maxRetryCount: 10,
    userAgentPrefix: 'snyk-api-ts-client',
});
class Orgs {
    constructor(fullResponse = false) {
        this.fullResponse = false;
        this.currentContext = {};
        this.fullResponse = fullResponse || false;
        const thisProperties = Object.getOwnPropertyNames(this);
        thisProperties.forEach((thisProperty) => {
            Object(this.currentContext)[thisProperty] = Object(this)[thisProperty];
        });
    }
    async get() {
        let url = '';
        let urlQueryParams = [];
        url = `/orgs`;
        if (urlQueryParams.length > 0) {
            url += `?${urlQueryParams.join('&')}`;
        }
        try {
            const result = await requestManager.request({ verb: 'get', url: url });
            if (!Object(this.currentContext)['fullResponse'] && result.data) {
                return result.data;
            }
            else {
                return result;
            }
        }
        catch (err) {
            throw new clientError_1.ClientError(err);
        }
    }
}
exports.Orgs = Orgs;
