// snyk-api-ts-client/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var group_1 = require("./client/generated/group");
Object.defineProperty(exports, "Group", { enumerable: true, get: function () { return group_1.Group; } });
var monitor_1 = require("./client/generated/monitor");
Object.defineProperty(exports, "Monitor", { enumerable: true, get: function () { return monitor_1.Monitor; } });
var org_1 = require("./client/generated/org");
Object.defineProperty(exports, "Org", { enumerable: true, get: function () { return org_1.Org; } });
var orgs_1 = require("./client/generated/orgs");
Object.defineProperty(exports, "Orgs", { enumerable: true, get: function () { return orgs_1.Orgs; } });
var reporting_1 = require("./client/generated/reporting");
Object.defineProperty(exports, "Reporting", { enumerable: true, get: function () { return reporting_1.Reporting; } });
var test_1 = require("./client/generated/test");
Object.defineProperty(exports, "Test", { enumerable: true, get: function () { return test_1.Test; } });
var user_1 = require("./client/generated/user");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return user_1.User; } });
var root_1 = require("./client/generated/root");
Object.defineProperty(exports, "Root", { enumerable: true, get: function () { return root_1.Root; } });
tslib_1.__exportStar(require("./client/utils/utils"), exports);
exports.GroupTypes = tslib_1.__importStar(require("./client/generated/group"));
exports.MonitorTypes = tslib_1.__importStar(require("./client/generated/monitor"));
exports.OrgTypes = tslib_1.__importStar(require("./client/generated/org"));
exports.OrgsTypes = tslib_1.__importStar(require("./client/generated/orgs"));
exports.ReportingTypes = tslib_1.__importStar(require("./client/generated/reporting"));
exports.TestTypes = tslib_1.__importStar(require("./client/generated/test"));
exports.UserTypes = tslib_1.__importStar(require("./client/generated/user"));
exports.RootTypes = tslib_1.__importStar(require("./client/generated/root"));
