// snyk-docker-plugin/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ImageType;
(function (ImageType) {
    ImageType[ImageType["Identifier"] = 0] = "Identifier";
    ImageType["DockerArchive"] = "docker-archive";
})(ImageType = exports.ImageType || (exports.ImageType = {}));
var OsReleaseFilePath;
(function (OsReleaseFilePath) {
    OsReleaseFilePath["Linux"] = "/etc/os-release";
    OsReleaseFilePath["LinuxFallback"] = "/usr/lib/os-release";
    OsReleaseFilePath["Lsb"] = "/etc/lsb-release";
    OsReleaseFilePath["Debian"] = "/etc/debian_version";
    OsReleaseFilePath["Alpine"] = "/etc/alpine-release";
    OsReleaseFilePath["RedHat"] = "/etc/redhat-release";
    OsReleaseFilePath["Oracle"] = "/etc/oracle-release";
})(OsReleaseFilePath = exports.OsReleaseFilePath || (exports.OsReleaseFilePath = {}));
// export type SupportedPackageManagers =
//   | 'rubygems'
//   | 'npm'
//   | 'yarn'
//   | 'maven'
//   | 'pip'
//   | 'sbt'
//   | 'gradle'
//   | 'golangdep'
//   | 'govendor'
//   | 'gomodules'
//   | 'nuget'
//   | 'paket'
//   | 'composer'
//   | 'cocoapods';
// just for reference
// export interface PluginMetadata {
//   name: string;
//   runtime?: string;
//   targetFile?: string;
//   packageManager?: SupportedPackageManagers;
//   // Per-plugin custom metadata
//   meta?: {
//     allSubProjectNames?: string[],
//     versionBuildInfo?: VersionBuildInfo,
//   };
//   // Docker-related fields
//   dockerImageId?: any;
//   imageLayers?: any;
//   packageFormatVersion?: string;
// }
